hydra/utils.py

```diff
@@ -53,67 +104,184 @@ def instantiate(config: Any, *args: Any, **kwargs: Any) -> Any:
+                    for item in config._iter_ex(resolve=True)
+                ]
+
+            # Otherwise, new OmegaConf container
+            lst = OmegaConf.create([], flags={"allow_objects": True})
+            lst._set_parent(config)
+            lst.extend(
+                [
+                    instantiate(item, parse=parse, convert=convert, recursive=recursive)
+                    for item in config._iter_ex(resolve=False)
+                ]
             )
-            return target(*args, **final_kwargs)
-        elif convert == ConvertMode.ALL:
-            final_kwargs = OmegaConf.to_container(final_kwargs, resolve=True)
-            return target(*args, **final_kwargs)
-        elif convert == ConvertMode.NONE:
-            return target(*args, **final_kwargs)
+            lst._metadata.object_type = config._metadata.object_type
+            return lst
+
         else:
-            assert False
-    except Exception as e:
-        # preserve the original exception backtrace
-        raise type(e)(
-            f"Error instantiating/calling '{_convert_target_to_string(target)}' : {e}"
-        ).with_traceback(sys.exc_info()[2])
+            if convert == ConvertMode.ALL:
+                return OmegaConf.to_container(config, resolve=True)
+            if convert == ConvertMode.PARTIAL:
+                return OmegaConf.to_container(
+                    config, resolve=True, exclude_structured_configs=True
+                )
+            return config
+
+    if isinstance(config, dict):
+        target = (
+            _resolve_target(config.pop(Keys.TARGET)) if _is_target(config) else None
+        )
+        if recursive:
+            config = {
+                key: instantiate(
+                    value, parse=parse, convert=convert, recursive=recursive
+                )
+                for key, value in config.items()
+            }
+        return _call_target(target, *args, **config) if target else config
+
+    if OmegaConf.is_dict(config):
+        if _is_target(config):
+            target = _resolve_target(config.pop(Keys.TARGET))
+            if recursive:
+                config = {
+                    key: instantiate(
+                        value, parse=parse, convert=convert, recursive=recursive
+                    )
+                    for key, value in config.items_ex(resolve=True)
+                }
+            return _call_target(target, *args, **config)
+
+        if recursive:
+            # If ALL or PARTIAL non structured, instantiate in dict
+            if convert == ConvertMode.ALL or (
+                convert == ConvertMode.PARTIAL and config._metadata.object_type is None
+            ):
+                return {
+                    key: instantiate(
+                        value, parse=parse, convert=convert, recursive=recursive
+                    )
+                    for key, value in config.items_ex(resolve=True)
+                }
```

Comment:
I think dict/list comprehension is no longer elegant once it starts to get this complex.
can you switch to a simple loop populating a `ret = {}` object?

Reply:
switched all multi-line comprehensions to standard assignments